src/components/actions/executeAction.ts

```diff
@@ -0,0 +1,98 @@
+import { ActionLibrary, ActionPayload, ActionQueueExecutePayload } from '../../types'
+import evaluateExpression from '@openmsupply/expression-evaluator'
+import functions from './evaluatorFunctions'
+import DBConnect from '../databaseConnect'
+import fetch from 'node-fetch'
+import { EvaluatorNode } from '@openmsupply/expression-evaluator/lib/types'
+import { getApplicationData } from './getApplicationData'
+import { ActionQueueStatus } from '../../generated/graphql'
+import config from '../../config'
+import { evaluateParameters } from './helpers'
+
+// Dev config
+const showApplicationDataLog = false
+
+const graphQLEndpoint = config.graphQLendpoint
+
+export async function executeAction(
+  payload: ActionPayload,
+  actionLibrary: ActionLibrary,
+  additionalObjects: any = {}
+): Promise<ActionQueueExecutePayload> {
+  // Get fresh applicationData for each Action
+  const applicationData = await getApplicationData({ payload })
+
+  // Debug helper console.log to inspect applicationData:
+  if (showApplicationDataLog) console.log('ApplicationData: ', applicationData)
+
+  const evaluatorParams = {
+    objects: { applicationData, functions, ...additionalObjects },
+    pgConnection: DBConnect,
+    APIfetch: fetch,
+    graphQLConnection: { fetch, endpoint: graphQLEndpoint },
+  }
+
+  // Evaluate condition
+  let condition
+  try {
+    condition = await evaluateExpression(
+      payload.condition_expression as EvaluatorNode,
+      evaluatorParams
+    )
+  } catch (err) {
+    console.log('>> Error evaluating condition for action:', payload.code)
+    return await DBConnect.executedActionStatusUpdate({
```

Comment:
hmm since you are refactoring this, passing through payload through to return of db method is not that clear. I think it's better in two steps:

```js
let actionResult = {
      status: ActionQueueStatus.Fail,
      error_log: 'Problem evaluating condition: ' + err,
      parameters_evaluated: null,
      output: null,
      id: payload.id,
    };

 await DBConnect.executedActionStatusUpdate(actionResult)
 return actionResult
```

Reply:
I think the reason I did it like that (it's been a while!) was so that if there is a database error, then that will get thrown back up to the caller instead of the action result. 

But I guess the database method would throw an error which would still get thrown from the catch block, wouldn't it?

So yeah, okay, will simplify:  918b9ba7

(BTW -- I seem to remember picking up this `return await` syntax from you in the first place 😉 ):
https://github.com/openmsupply/conforma-server/blob/47b013ede13d29ca10f3f71a03e911191245dfc9/database/updateRowPolicies.ts#L9-L11

Reply:
return await syntax is fine imo, in the signPromise it return a new thing, but in the executeAction, when return await executedActionStatusUpdate, it's just passing through the parameter it received, which cause a little bit confusion for me.

> But I guess the database method would throw an error which would still get thrown from the catch block, wouldn't it?

Yep, it would propagate either way